aws/resource_aws_codepipeline_webhook.go

```diff
@@ -0,0 +1,336 @@
+		},
+	}
+}
+
+func extractCodePipelineWebhookAttr(d *schema.ResourceData, attr string) (map[string]interface{}, error) {
+	if v, ok := d.GetOk(attr); ok {
+		l := v.([]interface{})
+		if len(l) <= 0 {
+			return nil, fmt.Errorf("Attribute %s is missing", attr)
+		}
+
+		data := l[0].(map[string]interface{})
+		return data, nil
+	}
+
+	return nil, fmt.Errorf("Could not find attribute %s", attr)
+}
+
+func extractCodePipelineWebhookRules(filters *schema.Set) ([]*codepipeline.WebhookFilterRule, error) {
+	var rules []*codepipeline.WebhookFilterRule
+
+	for _, f := range filters.List() {
+		r := f.(map[string]interface{})
+		filter := codepipeline.WebhookFilterRule{
+			JsonPath:    aws.String(r["json_path"].(string)),
+			MatchEquals: aws.String(r["match_equals"].(string)),
+		}
+
+		rules = append(rules, &filter)
+	}
+
+	return rules, nil
+}
+
+func extractCodePipelineWebhookAuthConfig(authType string, authConfig map[string]interface{}) (*codepipeline.WebhookAuthConfiguration, error) {
+	var conf codepipeline.WebhookAuthConfiguration
+	switch authType {
+	case codepipeline.WebhookAuthenticationTypeIp:
+		ipRange := authConfig["allowed_ip_range"].(string)
+		if ipRange == "" {
+			return nil, fmt.Errorf("An IP range must be set when using IP-based auth")
+		}
+
+		conf.AllowedIPRange = &ipRange
+
+		break
+	case codepipeline.WebhookAuthenticationTypeGithubHmac:
+		secretToken := authConfig["secret_token"].(string)
+		if secretToken == "" {
+			return nil, fmt.Errorf("Secret token must be set when using GITHUB_HMAC")
+		}
+
+		conf.SecretToken = &secretToken
+		break
+	case codepipeline.WebhookAuthenticationTypeUnauthenticated:
+		break
+	default:
+		return nil, fmt.Errorf("Invalid authentication type %s", authType)
+	}
+
+	return &conf, nil
+}
+
+func resourceAwsCodePipelineWebhookCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codepipelineconn
+	authType := d.Get("authentication").(string)
+
+	authConfig, err := extractCodePipelineWebhookAttr(d, "authentication_configuration")
+	if err != nil {
+		return err
+	}
+
+	rules, err := extractCodePipelineWebhookRules(d.Get("filter").(*schema.Set))
+	if err != nil {
+		return err
+	}
+
+	conf, err := extractCodePipelineWebhookAuthConfig(authType, authConfig)
+	if err != nil {
+		return err
+	}
+
+	request := &codepipeline.PutWebhookInput{
+		Webhook: &codepipeline.WebhookDefinition{
+			Authentication:              aws.String(authType),
+			Filters:                     rules,
+			Name:                        aws.String(d.Get("name").(string)),
+			TargetAction:                aws.String(d.Get("target_action").(string)),
+			TargetPipeline:              aws.String(d.Get("target_pipeline").(string)),
+			AuthenticationConfiguration: conf,
+		},
+	}
+
+	webhook, err := conn.PutWebhook(request)
+	if err != nil {
+		return fmt.Errorf("Error creating webhook: %s", err)
+	}
+
+	arn := *webhook.Webhook.Arn
+	d.SetId(arn)
+
+	url := *webhook.Webhook.Url
+	d.Set("url", url)
```

Comment:
`d.Set()` should only be called in the `Read` function, unless this information is not available then. 👍